File: backend/routers/reactions.py
"""API endpoints for message reactions."""

from __future__ import annotations

import logging
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy import select, delete
from sqlalchemy.ext.asyncio import AsyncSession

from ..core.database import get_db
from ..models.server import Message, Reaction
from .deps import CurrentUser
from .schemas import ReactionGroupResponse, ReactionUserResponse

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/messages/{message_id}/reactions", tags=["reactions"])


class ReactionCreate(BaseModel):
    emoji: str


class ReactionGroupWithReacted(ReactionGroupResponse):
    """Grouped reactions with current user's reaction status."""

    reacted: bool  # Whether current user has reacted with this emoji


class ReactionResponse(BaseModel):
    id: int
    message_id: int
    user_id: int
    username: str
    emoji: str
    created_at: datetime

    class Config:
        from_attributes = True


def group_reactions(
    reactions: list[Reaction], current_user_id: int
) -> list[ReactionGroupWithReacted]:
    """Group reactions by emoji and include user info."""
    groups: dict[str, ReactionGroupWithReacted] = {}

    for r in reactions:
        if r.emoji not in groups:
            groups[r.emoji] = ReactionGroupWithReacted(
                emoji=r.emoji, count=0, users=[], reacted=False
            )
        groups[r.emoji].count += 1
        groups[r.emoji].users.append(
            ReactionUserResponse(id=r.user_id, username=r.username)
        )
        if r.user_id == current_user_id:
            groups[r.emoji].reacted = True

    return list(groups.values())


@router.post("", response_model=ReactionResponse, status_code=status.HTTP_201_CREATED)
async def add_reaction(
    message_id: int,
    payload: ReactionCreate,
    user: CurrentUser,
    db: AsyncSession = Depends(get_db),
):
    """Add a reaction to a message. If already reacted with same emoji, returns existing."""
    # Validate emoji (basic check - non-empty, reasonable length)
    emoji = payload.emoji.strip()
    if not emoji or len(emoji) > 32:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid emoji",
        )

    # Verify message exists and is not deleted
    msg_result = await db.execute(select(Message).where(Message.id == message_id))
    message = msg_result.scalar_one_or_none()
    if not message:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Message not found"
        )
    if message.is_deleted:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Cannot react to deleted message",
        )

    # Check if user already reacted with this emoji
    existing_result = await db.execute(
        select(Reaction).where(
            Reaction.message_id == message_id,
            Reaction.user_id == user["id"],
            Reaction.emoji == emoji,
        )
    )
    existing = existing_result.scalar_one_or_none()
    if existing:
        # Already reacted, return existing
        return existing

    # Create new reaction
    reaction = Reaction(
        message_id=message_id,
        user_id=user["id"],
        username=user.get("nickname") or user["username"],
        emoji=emoji,
    )
    db.add(reaction)
    await db.commit()
    await db.refresh(reaction)

    # Broadcast reaction added via WebSocket
    from ..websocket.manager import chat_manager

    await chat_manager.broadcast_to_all_users(
        {
            "type": "reaction_added",
            "message_id": message_id,
            "channel_id": message.channel_id,
            "emoji": emoji,
            "user_id": user["id"],
            "username": user.get("nickname") or user["username"],
        },
    )

    return reaction


@router.delete("/{emoji}", status_code=status.HTTP_204_NO_CONTENT)
async def remove_reaction(
    message_id: int,
    emoji: str,
    user: CurrentUser,
    db: AsyncSession = Depends(get_db),
):
    """Remove a reaction from a message."""
    # URL decode emoji (in case it's encoded)
    from urllib.parse import unquote

    emoji = unquote(emoji)

    # Verify message exists
    msg_result = await db.execute(select(Message).where(Message.id == message_id))
    message = msg_result.scalar_one_or_none()
    if not message:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Message not found"
        )

    # Find and delete the reaction
    result = await db.execute(
        select(Reaction).where(
            Reaction.message_id == message_id,
            Reaction.user_id == user["id"],
            Reaction.emoji == emoji,
        )
    )
    reaction = result.scalar_one_or_none()
    if not reaction:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Reaction not found"
        )

    await db.delete(reaction)
    await db.commit()

    # Broadcast reaction removed via WebSocket
    from ..websocket.manager import chat_manager

    await chat_manager.broadcast_to_all_users(
        {
            "type": "reaction_removed",
            "message_id": message_id,
            "channel_id": message.channel_id,
            "emoji": emoji,
            "user_id": user["id"],
        },
    )

    return None


@router.get("", response_model=list[ReactionGroupWithReacted])
async def get_reactions(
    message_id: int,
    user: CurrentUser,
    db: AsyncSession = Depends(get_db),
):
    """Get all reactions for a message, grouped by emoji."""
    # Verify message exists
    msg_result = await db.execute(select(Message).where(Message.id == message_id))
    message = msg_result.scalar_one_or_none()
    if not message:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Message not found"
        )

    # Get all reactions for this message
    result = await db.execute(
        select(Reaction)
        .where(Reaction.message_id == message_id)
        .order_by(Reaction.created_at)
    )
    reactions = result.scalars().all()

    return group_reactions(list(reactions), user["id"])

File: backend-go/internal/handler/reactions.go
package handler

import (
	"database/sql"
	"net/http"
	"net/url"
	"strconv"

	"github.com/labstack/echo/v4"

	"github.com/RMS-Server/rms-discord-go/internal/middleware"
	"github.com/RMS-Server/rms-discord-go/internal/sso"
)

// ReactionHandler handles message reaction endpoints.
type ReactionHandler struct {
	db  *sql.DB
	sso *sso.Client
}

func NewReactionHandler(db *sql.DB, sso *sso.Client) *ReactionHandler {
	return &ReactionHandler{db: db, sso: sso}
}

type reactionCreateReq struct {
	Emoji string `json:"emoji"`
}

type reactionResp struct {
	ID        int64  `json:"id"`
	MessageID int64  `json:"message_id"`
	UserID    int64  `json:"user_id"`
	Username  string `json:"username"`
	Emoji     string `json:"emoji"`
	CreatedAt string `json:"created_at"`
}

type reactionGroupWithReacted struct {
	Emoji   string             `json:"emoji"`
	Count   int                `json:"count"`
	Users   []reactionUserResp `json:"users"`
	Reacted bool               `json:"reacted"`
}

// AddReaction adds a reaction to a message (idempotent).
// POST /api/messages/:message_id/reactions
func (h *ReactionHandler) AddReaction(c echo.Context) error {
	user := middleware.GetUser(c)
	messageID, err := strconv.ParseInt(c.Param("message_id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid message id"})
	}

	var req reactionCreateReq
	if err := c.Bind(&req); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid request body"})
	}
	emoji := req.Emoji
	if emoji == "" || len(emoji) > 32 {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid emoji"})
	}

	// Verify message exists and not deleted
	var isDeleted bool
	var channelID int64
	err = h.db.QueryRow("SELECT channel_id, is_deleted FROM messages WHERE id = ?", messageID).Scan(&channelID, &isDeleted)
	if err == sql.ErrNoRows {
		return c.JSON(http.StatusNotFound, map[string]string{"error": "message not found"})
	}
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	if isDeleted {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "cannot react to deleted message"})
	}

	// Check existing
	var existing reactionResp
	err = h.db.QueryRow(
		"SELECT id, message_id, user_id, username, emoji, created_at FROM reactions WHERE message_id = ? AND user_id = ? AND emoji = ?",
		messageID, user.ID, emoji,
	).Scan(&existing.ID, &existing.MessageID, &existing.UserID, &existing.Username, &existing.Emoji, &existing.CreatedAt)
	if err == nil {
		return c.JSON(http.StatusCreated, existing)
	}

	username := user.Nickname
	if username == "" {
		username = user.Username
	}

	// INSERT IGNORE for idempotency
	h.db.Exec(
		"INSERT IGNORE INTO reactions (message_id, user_id, username, emoji) VALUES (?, ?, ?, ?)",
		messageID, user.ID, username, emoji,
	)

	// Fetch the created reaction
	var resp reactionResp
	h.db.QueryRow(
		"SELECT id, message_id, user_id, username, emoji, created_at FROM reactions WHERE message_id = ? AND user_id = ? AND emoji = ?",
		messageID, user.ID, emoji,
	).Scan(&resp.ID, &resp.MessageID, &resp.UserID, &resp.Username, &resp.Emoji, &resp.CreatedAt)

	// Broadcast via WS
	if BroadcastFunc != nil {
		BroadcastFunc(channelID, map[string]interface{}{
			"type":       "reaction_added",
			"message_id": messageID,
			"channel_id": channelID,
			"emoji":      emoji,
			"user_id":    user.ID,
			"username":   username,
		})
	}

	return c.JSON(http.StatusCreated, resp)
}

// RemoveReaction removes a reaction (URL-decode emoji).
// DELETE /api/messages/:message_id/reactions/:emoji
func (h *ReactionHandler) RemoveReaction(c echo.Context) error {
	user := middleware.GetUser(c)
	messageID, err := strconv.ParseInt(c.Param("message_id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid message id"})
	}

	emoji, err := url.PathUnescape(c.Param("emoji"))
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid emoji"})
	}

	// Verify message
	var channelID int64
	err = h.db.QueryRow("SELECT channel_id FROM messages WHERE id = ?", messageID).Scan(&channelID)
	if err == sql.ErrNoRows {
		return c.JSON(http.StatusNotFound, map[string]string{"error": "message not found"})
	}
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	res, err := h.db.Exec(
		"DELETE FROM reactions WHERE message_id = ? AND user_id = ? AND emoji = ?",
		messageID, user.ID, emoji,
	)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	n, _ := res.RowsAffected()
	if n == 0 {
		return c.JSON(http.StatusNotFound, map[string]string{"error": "reaction not found"})
	}

	if BroadcastFunc != nil {
		BroadcastFunc(channelID, map[string]interface{}{
			"type":       "reaction_removed",
			"message_id": messageID,
			"channel_id": channelID,
			"emoji":      emoji,
			"user_id":    user.ID,
		})
	}

	return c.NoContent(http.StatusNoContent)
}

// GetReactions returns reactions grouped by emoji.
// GET /api/messages/:message_id/reactions
func (h *ReactionHandler) GetReactions(c echo.Context) error {
	user := middleware.GetUser(c)
	messageID, err := strconv.ParseInt(c.Param("message_id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid message id"})
	}

	var exists int
	if err := h.db.QueryRow("SELECT 1 FROM messages WHERE id = ?", messageID).Scan(&exists); err != nil {
		return c.JSON(http.StatusNotFound, map[string]string{"error": "message not found"})
	}

	rows, err := h.db.Query(
		"SELECT emoji, user_id, username FROM reactions WHERE message_id = ? ORDER BY created_at", messageID,
	)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	defer rows.Close()

	groupMap := map[string]*reactionGroupWithReacted{}
	var order []string
	for rows.Next() {
		var emoji, username string
		var userID int64
		rows.Scan(&emoji, &userID, &username)
		if _, ok := groupMap[emoji]; !ok {
			groupMap[emoji] = &reactionGroupWithReacted{Emoji: emoji}
			order = append(order, emoji)
		}
		g := groupMap[emoji]
		g.Count++
		g.Users = append(g.Users, reactionUserResp{ID: userID, Username: username})
		if userID == int64(user.ID) {
			g.Reacted = true
		}
	}

	result := make([]reactionGroupWithReacted, 0, len(order))
	for _, e := range order {
		result = append(result, *groupMap[e])
	}
	return c.JSON(http.StatusOK, result)
}
